pkg/kubelet/winstats/winstats.go

```diff
@@ -0,0 +1,125 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package winstats provides a client to get node and pod level stats on windows
+package winstats
+
+import (
+	"time"
+
+	cadvisorapi "github.com/google/cadvisor/info/v1"
+	cadvisorapiv2 "github.com/google/cadvisor/info/v2"
+)
+
+// Client is an object that is used to get stats information.
+type Client struct {
+	winNodeStatsClient
+}
+
+type winNodeStatsClient interface {
+	startMonitoring() error
+	getNodeStats() (nodeStats, error)
+	getMachineInfo() (*cadvisorapi.MachineInfo, error)
+	getVersionInfo() (*cadvisorapi.VersionInfo, error)
+}
+
+type metric struct {
+	Name      string
+	Value     uint64
+	Timestamp time.Time
+}
+
+type nodeStats struct {
+	cpuUsageCoreNanoSeconds     metric
+	memoryPrivWorkingSetBytes   metric
+	memoryCommittedBytes        metric
+	memoryPhysicalCapacityBytes uint64
+	kernelVersion               string
+}
+
+// NewClient constructs a Client.
+func NewClient(statsClient winNodeStatsClient) (*Client, error) {
+	client := new(Client)
+	client.winNodeStatsClient = statsClient
+
+	err := client.startMonitoring()
+	if err != nil {
+		return nil, err
+	}
+
+	return client, nil
+}
+
+// WinContainerInfos returns a map of container infos. The map contains node and
+// pod level stats. Analogous to cadvisor GetContainerInfoV2 method.
+func (c *Client) WinContainerInfos() (map[string]cadvisorapiv2.ContainerInfo, error) {
+	infos := make(map[string]cadvisorapiv2.ContainerInfo)
+	rootContainerInfo, err := c.createRootContainerInfo()
+	if err != nil {
+		return nil, err
+	}
+
+	infos["/"] = *rootContainerInfo
+
+	return infos, nil
+}
+
+// WinMachineInfo returns a cadvisorapi.MachineInfo with details about the
+// node machine. Analogous to cadvisor MachineInfo method.
+func (c *Client) WinMachineInfo() (*cadvisorapi.MachineInfo, error) {
+	return c.getMachineInfo()
+}
+
+// WinVersionInfo returns a  cadvisorapi.VersionInfo with version info of
+// the kernel and docker runtime. Analogous to cadvisor VersionInfo method.
+func (c *Client) WinVersionInfo() (*cadvisorapi.VersionInfo, error) {
+	return c.getVersionInfo()
+}
+
+func (c *Client) createRootContainerInfo() (*cadvisorapiv2.ContainerInfo, error) {
+	nodeStats, err := c.getNodeStats()
+
+	if err != nil {
+		return nil, err
+	}
+	var stats []*cadvisorapiv2.ContainerStats
+
+	stats = append(stats, &cadvisorapiv2.ContainerStats{
+		Timestamp: nodeStats.cpuUsageCoreNanoSeconds.Timestamp,
```

Comment:
Here we use the timestamp of the CPU stats for memory stats, which seems incorrect. The timestamp for the two memory stats are unused? How about using a single goroutine to get all the three metrics?

Reply:
The metrics come over a channel every second. I believe the timestamp here is for the overall stats, so I used cpu time stamp. To make it a little bit more clear, I changed this by adding a `lastUpdatedTime` to node stats and using that timestamp instead.

Reply:
In this way, the `Timestamp` in `metric` is not used. More importantly, since we are using different goroutine to get the metrics, it's possible that we get the CPU stats and memory stats updates at different intervals and therefore the last timestamp is inaccurate.

How about using a single goroutine to get all three metrics? To do that, we can change the `metric` to something like this
```
type metrics struct {
    cpuUsageCoreNanoSeconds uint64
    memoryPrivWorkingSetBytes uint64
    memoryCommittedBytes uint64
    Timestamp time.Time
}
```
And in one goroutine call `readPerformanceCounter` to populate the 3 metrics one by one? This seems to be clearer and can save goroutines too.

Reply:
That can be done, but either way if the three metrics are queried one by one (with delay in between), the timestamp will not be sequential. The timestamp will need to be either at the beginning of the queries or at the end. I think using goroutines are suppose to be cheap (I'm only making three of them here...).

Reply:
But they are unnecessary, and you'd have to document that the three goroutines must have the same interval. I think using one goroutine to update the three metrics at once is simpler and more efficient. This seems to be also what @yujuhong mentioned in https://github.com/kubernetes/kubernetes/pull/50396/files#r138158481.